Add support for `EXTENSIONS_LOCATION` setting

rijkvanzanten opened the pull request:
## Scope

Allows you to chose the storage location from which extensions are loaded. The API does this by copying over the extensions to a local folder (`./.directus` by default, configurable through `$TEMP_PATH`) on extensions load, and running the extensions from that folder. This syncing is required to allow us to load in the extensions as regular node modules like before.

What's changed:

- Added an `EXTENSIONS_LOCATION` env var
- If set, the API will copy over the extensions from the `EXTENSIONS_PATH` in the storage location to a local tmp folder 
- Added the `TEMP_PATH` env var to configure the path of the local tmp folder

## Potential Risks / Drawbacks

- It syncs to a local folder before startup, and then uses the local folder as if it was just a regular extensions folder. This should in theory mean that nothing really changed in how extensions operate outside of downloading the extensions during startup.
- If your extensions folder has tons of files, for example when you put `src` and `node_modules` in the extension folder, this will slow down the startup process quite a bit

## Review Notes / Questions

- I thought about checking out the extensions manifest to find what files we need to download to run the extension, but realized that API extensions can technically have their own dependency chain, making that a risky endeavor

---

Fixes #19803

rijkvanzanten committed:
Add EXTENSIONS_LOCATION env var

rijkvanzanten committed:
Install promise-queue

rijkvanzanten committed:
Git ignore .directus temp folder

rijkvanzanten committed:
Add TEMP_PATH env var

rijkvanzanten committed:
Add util to get the local fs path to extensions

rijkvanzanten committed:
Add util to sync extensions from storage location to local

rijkvanzanten committed:
Sync extensions from location on load

rijkvanzanten committed:
Use extensions path util

rijkvanzanten committed:
Add changeset

rijkvanzanten committed:
Add reference

rijkvanzanten committed:
Abbreviation kills

paescuj commented on api/src/extensions/lib/sync-extensions.ts:
```diff
@@ -0,0 +1,41 @@
+		const writeStream = createWriteStream(destPath);
+
+		await queue.add(() => pipeline(readStream, writeStream));
+	}
```
Yeah, I think right now it runs sequentially and don't make real usage of `p-queue`. We shouldn't await `queue.add` but `queue.onIdle()` at the end.

However, with that we must ensure errors are caught correctly: https://github.com/sindresorhus/p-queue#addfn-options

DanielBiegler replied:
Can confirm! Just tried it out to scratch my curiosity itch and we really shouldnt await adding.

```js
import Queue from 'p-queue';

const queue = new Queue({ concurrency: 2 })
const time = Date.now();

queue.add(async () => new Promise((res, rej) => { setTimeout(() => {
  console.log("PROMISE RAN - ELAPSED:", Date.now() - time, "ms"); res();
}, 2000); }))

queue.add(async () => new Promise((res, rej) => { setTimeout(() => {
  console.log("PROMISE RAN - ELAPSED:", Date.now() - time, "ms"); res();
}, 2000); }))

queue.add(async () => new Promise((res, rej) => { setTimeout(() => {
  console.log("PROMISE RAN - ELAPSED:", Date.now() - time, "ms"); res();
}, 2000); }))

queue.add(async () => new Promise((res, rej) => { setTimeout(() => {
  console.log("PROMISE RAN - ELAPSED:", Date.now() - time, "ms"); res();
}, 2000); }))

queue.add(async () => new Promise((res, rej) => { setTimeout(() => {
  console.log("PROMISE RAN - ELAPSED:", Date.now() - time, "ms"); res();
}, 2000); }))

console.log("BEFORE IDLE")

await queue.onIdle()

console.log("AFTER IDLE: WE ARE DONE")
```

Correctly results in:

```
BEFORE IDLE
PROMISE RAN - ELAPSED: 2004 ms
PROMISE RAN - ELAPSED: 2004 ms
PROMISE RAN - ELAPSED: 4007 ms
PROMISE RAN - ELAPSED: 4008 ms
PROMISE RAN - ELAPSED: 6008 ms
AFTER IDLE: WE ARE DONE
```

Whereas awaiting the adding results in:

```
PROMISE RAN - ELAPSED: 2005 ms
PROMISE RAN - ELAPSED: 4011 ms
PROMISE RAN - ELAPSED: 6012 ms
PROMISE RAN - ELAPSED: 8015 ms
PROMISE RAN - ELAPSED: 10016 ms
BEFORE IDLE
AFTER IDLE: WE ARE DONE
```

licitdev reviewed (commented):
As the `.directus` folder is shared between instances when running with multiple PM2 instances, I think there might be a chance of file errors due to write locks, but I haven't ran into it yet in my tests. The files just gets rewritten over and over again.

Should we implement some form of hash check to skip writing files that already exist? However, it will not eliminate the write locks issue entirely. 🤔

rijkvanzanten commented:
> As the .directus folder is shared between instances when running with multiple PM2 instances, I think there might be a chance of file errors due to write locks, but I haven't ran into it yet in my tests. The files just gets rewritten over and over again.

Good one 👍🏻 

I think what we should do is use a Redis-lock similar to schedule flows that's scoped to the machine id (/ docker container id) to make sure only one of the processes handles the download and syncing 🙂

rijkvanzanten committed:
Default to ./node_modules/.directus

rijkvanzanten committed:
Fix copying issue if path starts relative

rijkvanzanten committed:
Don't load extensions simultaneously

rijkvanzanten committed:
Add sync wait in other process before continuing

rijkvanzanten committed:
Don't await queue

h/t Daniel

rijkvanzanten committed:
Await queue completion

rijkvanzanten committed:
Run formatter

rijkvanzanten committed:
Use status file instead of lock

rijkvanzanten committed:
Only load status file if it exists

rijkvanzanten committed:
Fix async usage of mid

licitdev committed:
Fix relative paths such as ../ ./ . in EXTENSIONS_PATH

licitdev committed:
Skip syncing if already DONE

licitdev commented on api/src/extensions/manager.ts:
```diff
@@ -173,7 +175,8 @@ export class ExtensionManager {
 		try {
-			await ensureExtensionDirs(env['EXTENSIONS_PATH'], NESTED_EXTENSION_TYPES);
+			await syncExtensions();
+			await ensureExtensionDirs(getExtensionsPath(), NESTED_EXTENSION_TYPES);
```
Can we shift `ensureExtensionDirs()` into `syncExtensions()` so that only the primary process creates the required directories during the sync?

rijkvanzanten replied:
We can!

licitdev replied:
Updated. Thanks! 👍

licitdev committed:
Docs for interaction with EXTENSIONS_AUTO_RELOAD

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

licitdev committed:
Run ensureExtensionDirs() for primary process only

licitdev commented on api/src/utils/validate-storage.ts:
```diff
@@ -29,8 +30,8 @@ export async function validateStorage(): Promise<void> {
+		await access(getExtensionsPath(), constants.R_OK);
 	} catch {
-		logger.warn(`Extensions directory (${path.resolve(env['EXTENSIONS_PATH'])}) is not readable!`);
+		logger.warn(`Extensions directory (${path.resolve(getExtensionsPath())}) is not readable!`);
```
When running locally for the first time, this warning will surface as the temp extension folder hasn't been created yet.

Should we add a check here on whether `EXTENSIONS_LOCATION` is set?

br41nslug replied:
I agree, something along these lines would be in order i think 🤔 
```suggestion
        if (env['EXTENSIONS_LOCATION'] === 'local') {
		try {
			await access(getExtensionsPath(), constants.R_OK);
		} catch {
			logger.warn(`Extensions directory (${path.resolve(getExtensionsPath())}) is not readable!`);
		}
	}
```

licitdev replied:
Thanks for the heads up, patched via https://github.com/directus/directus/pull/20207/commits/e47066604ba50e0bf6f02a7175808b25dc05fab5.

I think it should be bypassed for all, when `EXTENSIONS_LOCATION` is set, as the storage driver name can be any string including `local`.

The path is validated when the temp extensions folder is created during the sync.
https://github.com/directus/directus/blob/e47066604ba50e0bf6f02a7175808b25dc05fab5/api/src/extensions/lib/sync-extensions.ts#L47-L48

licitdev committed:
Remove accidental typo

licitdev committed:
Perform a fresh sync if container restarts

licitdev committed:
Fix write error on top level instead of using recursive in writeFile

licitdev commented on api/src/extensions/lib/sync-extensions.ts:
```diff
@@ -0,0 +1,94 @@
+
+	// Make sure we don't overload the file handles
+	const queue = new Queue({ concurrency: 1000 });
+
+	for await (const filepath of disk.list(env['EXTENSIONS_PATH'])) {
+		const readStream = await disk.read(filepath);
+
+		// We want files to be stored in the root of `$TEMP_PATH/extensions`, so gotta remove the
+		// extensions path on disk from the start of the file path
+		const destPath = join(
+			getExtensionsPath(),
+			filepath.substring(resolve('/', normalize(env['EXTENSIONS_PATH'])).length - 1)
+		);
```
Can we extract a `prefixPathLength` to optimise the resolve and normalize calls? 

The naming is tricky, I had initially started off with `externalExtensionsPathLength` as this is the path from the "external" extensions storage.

```suggestion
	const prefixPathLength = resolve('/', normalize(env['EXTENSIONS_PATH'])).length - 1;

	for await (const filepath of disk.list(env['EXTENSIONS_PATH'])) {
		const readStream = await disk.read(filepath);

		// We want files to be stored in the root of `$TEMP_PATH/extensions`, so gotta remove the
		// extensions path on disk from the start of the file path
		const destPath = join(extensionsPath, filepath.substring(prefixPathLength));
```

br41nslug commented on api/src/extensions/lib/sync-extensions.ts:
```diff
@@ -0,0 +1,98 @@
+
+		// We want files to be stored in the root of `$TEMP_PATH/extensions`, so gotta remove the
+		// extensions path on disk from the start of the file path
+		const destPath = join(
+			extensionsPath,
+			filepath.substring(resolve('/', normalize(env['EXTENSIONS_PATH'])).length - 1)
+		);
```
This logic seems flawed as all my extension folders get created without the first character, resulting in them not being recognized as extensions and not loaded.

![image](https://github.com/directus/directus/assets/9389634/0c1c6139-b17d-467d-b46d-289ba134e703)

```suggestion
		const destPath = join(
			extensionsPath,
			filepath.substring(normalize(env['EXTENSIONS_PATH']).length)
		);
```

licitdev replied:
Great catch on the difference in file path conventions! I've just pushed https://github.com/directus/directus/pull/20207/commits/938330116f2a881ce31cd88b64b5bf3746444faa, does it resolve this issue? 🧐

br41nslug replied:
This solution does not work im afraid, the underlying issue lies with the use of `resolve()` not with the directory separator. Both directory separators are only 1 character in length but the difference is coming from the Drive letter that is added in `resolve` (hence my suggested change).
![image](https://github.com/directus/directus/assets/9389634/c5729292-3c36-47f9-bdae-a664a3281feb)

br41nslug replied:
```suggestion
		const destPath = join(
			extensionsPath,
			filepath.substring(resolve(sep, normalize('./extensions')).length - resolve(sep).length)
		);
```

licitdev replied:
@br41nslug Great idea! Pushed via https://github.com/directus/directus/pull/20207/commits/6e78b0ebea9d9274d000789fc971f071b7096aff. Thank you! 🙏

br41nslug reviewed (changes requested):
Ran into some possibly Windows specific issues 😇

licitdev commented on api/src/env.ts:
```diff
@@ -225,6 +227,8 @@ const defaults: Record<string, any> = {
 	MAX_BATCH_MUTATION: Infinity,
 	ROBOTS_TXT: 'User-agent: *\nDisallow: /',
 
+	TEMP_PATH: './node_modules/.directus',
```
Yup! Any thoughts on https://github.com/directus/directus/pull/20207#discussion_r1379165900?

licitdev committed:
Use path.sep

licitdev committed:
Merge branch 'main' into issue/19803

licitdev committed:
Remove obsolete getEnv

licitdev committed:
Merge branch 'main' into issue/19803

licitdev committed:
Fix substring path length for Windows

Co-authored-by: Brainslug <br41nslug@users.noreply.github.com>

paescuj committed:
Use latest p-queue version

br41nslug reviewed (approved):
The extensions now load as expected for me 🚀 I do still think we need to avoid throwing that "Extensions directory not readable" warning, since according to my tests it does work making the warning confusing.

And a strange thing i noticed but dont think it has to be a blocker, extensions are getting synced independ of the command that is being run. For example `npx directus lalala` will starting syncing the remote extension before realizing that `lalala` is not a known command.

paescuj commented on api/src/extensions/lib/sync-extensions.ts:
```diff
@@ -0,0 +1,94 @@
+	const messenger = getMessenger();
+
+	const isPrimaryProcess =
+		String(process.env['NODE_APP_INSTANCE']) === '0' || process.env['NODE_APP_INSTANCE'] === undefined;
```
Is this really required? Values in `process.env` should always be a string 🤔 

```suggestion
	process.env['NODE_APP_INSTANCE'] === '0' || process.env['NODE_APP_INSTANCE'] === undefined;
```

rijkvanzanten replied:
Technically not, but our `env.ts` handling does some magic in that global space so better safe than sorry here

paescuj commented on api/src/extensions/lib/sync-extensions.ts:
```diff
@@ -0,0 +1,94 @@
+		});
+	}
+
+	if (!env['EXTENSIONS_LOCATION']) {
+		return;
+	}
```
Shouldn't we move this check to the top?

rijkvanzanten replied:
Yeah 👍🏻 Remnant of when this file was shared with other extension loading biz

paescuj reviewed (changes requested):
I'm currently a bit worried about deviating states of extensions when using multiple instances across different hosts (horizontally scaled):
1. Two instances are started
2. Extensions are modified
3. One of the instances gets restarted for whatever reason and picks up the new state of the extensions while the other instance still runs on the old state

Therefore I think we should also have cross-host notification/subscription when extensions are (re-)synced.

**Edit:** I think we'll need that anyway for the Marketplace...

br41nslug commented on api/src/extensions/lib/sync-extensions.ts:
```diff
@@ -0,0 +1,94 @@
+
+	// Make sure we don't overload the file handles
+	const queue = new Queue({ concurrency: 1000 });
+	const resolvedPathPrefixLength = resolve(sep, normalize(env['EXTENSIONS_PATH'])).length - resolve(sep).length;
+
+	for await (const filepath of disk.list(env['EXTENSIONS_PATH'])) {
+		const readStream = await disk.read(filepath);
+
+		// We want files to be stored in the root of `$TEMP_PATH/extensions`, so gotta remove the
+		// extensions path on disk from the start of the file path
+		const destPath = join(extensionsPath, filepath.substring(resolvedPathPrefixLength));
```
Did a few tests on Windows and nicks suggestion works as intended for me 😄

paescuj commented on api/src/extensions/lib/sync-extensions.ts:
```diff
@@ -0,0 +1,94 @@
+		return;
+	}
+
+	const currentTime = new Date();
+	currentTime.setMilliseconds(0);
+	const systemStartTimestamp = new Date(currentTime.getTime() - uptime() * 1000).toISOString();
+	const isSyncRequired = (await getSyncTimestamp()) !== systemStartTimestamp;
+
+	if (!isSyncRequired) {
+		return;
+	}
```
I'd probably move that logic into the dedicated `sync-timestamp.ts` file

licitdev replied:
Done. Refactored in https://github.com/directus/directus/pull/20207/commits/59c9355bb6f7d49170991b979758be946f76f54f where this function is also used in `checkIsSyncRequired()`.

paescuj replied:
Thank you @licitdev ❣️❣️❣️

rijkvanzanten committed:
Update .gitignore

Co-authored-by: Pascal Jufer <pascal-jufer@bluewin.ch>

rijkvanzanten commented:
> Therefore I think we should also have cross-host notification/subscription when extensions are (re-)synced.

Edit: I think we'll need that anyway for the Marketplace...

@paescuj this is one of the todo's for marketplace, but I don't want to conflate it in this PR. As of right now, new extensions require a restart, so that's a workable situation for this particular PR. Once the rest of the marketplace work wraps up it will have to message across and sync on changes to your point 👍🏻

rijkvanzanten committed:
Move location check to top

rijkvanzanten committed:
Apply nicola's suggestion

rijkvanzanten committed:
Remove unused import

paescuj commented:
> @paescuj this is one of the todo's for marketplace, but I don't want to conflate it in this PR. As of right now, new extensions require a restart, so that's a workable situation for this particular PR. Once the rest of the marketplace work wraps up it will have to message across and sync on changes to your point 👍🏻

Okay, sounds good to me 🚀

licitdev committed:
Simplify to timestamp only as per pascal's suggestion

licitdev committed:
Validate access to extensions directory only if EXTENSIONS_LOCATION is set

licitdev committed:
Mention the copying of extensions in docs

licitdev committed:
Add clarification for EXTENSIONS_PATH

licitdev committed:
Improve clarity

br41nslug reviewed (approved):
Did a last check after the recent changes and while these are not blocking they may deserve a mention in the docs:
1. I find it confusing that there is a difference between `EXTENSIONS_LOCATION=""` and `EXTENSIONS_LOCATION="local"` where the first will do the default `"local"` behavior and load from the extension folder while the latter will look for the extensions folder inside the uploads folder
2. It can be non-obvious that the TEMP_PATH can absolutely not overlap either the uploads folder or the extensions folder (depending where you're loading your extensions from in 1.)

These are however just misconfiguration issues, the functionality itself works as expected 😄

licitdev committed:
Fix weird os.uptime() float

paescuj commented on docs/self-hosted/config-options.md:
```diff
@@ -898,19 +899,28 @@ const publicUrl = process.env.PUBLIC_URL;
 By default, extensions are not cached. The input data type for this environment variable is the same as
 [`CACHE_TTL`](#cache).
 
+<sup>[4]</sup> Extensions defaults to the local filesystem. When `EXTENSIONS_LOCATION` is configured, the extensions are
+copied into a local directory within `TEMP_PATH`.
```
I'd slightly reword that to something like
```suggestion
<sup>[4]</sup> By default extensions are loaded from the local filesystem. `EXTENSIONS_LOCATION` can be used to load extensions from a remote location instead.
```

licitdev replied:
Sure, feel free to refactor accordingly. 🙂 
 
The reasoning behind this footnote was to highlight the copying to the `TEMP_PATH` when `EXTENSIONS_LOCATION` is configured.

I think it partially helps to address point 1 of @br41nslug's review which mentioned that there's a difference between leaving it as default vs specifying `EXTENSIONS_LOCATION="local"`, where the latter would copy the files into an internal extensions folder.

licitdev committed:
Simplify to always sync on the first instance

paescuj committed:
Detailed explanation for `EXTENSIONS_LOCATION`

paescuj committed:
Fix dict ¯\_(ツ)_/¯

paescuj committed:
Run ensureExtensionDirs with no EXTENSIONS_LOCATION configured

paescuj committed:
Add note about multiple pm2 instances

paescuj committed:
Merge branch 'main' into issue/19803

licitdev committed:
Try fixing spell check with a space

licitdev committed:
Skip folders for list() in S3 driver

licitdev committed:
Fix invalid . path for S3

paescuj committed:
Clean-up patches

paescuj committed:
Remove env overrides

Those are remnants of earlier versions of the Dockerfile.
Explicit overrides were necessary since we used paths which deviated from
the defaults (./extensions | ./uploads).

Now, with this new feature it leads to unexpected behaviour as the docs
states that the EXTENSIONS_PATH defaults to './extensions' but this
isn't true inside Docker and extensions inside storage location wouldn't
be found.

paescuj committed:
Ignore txt files for linting

paescuj committed:
Ignore Dockerfile

paescuj committed:
Add comment for ensureExtensionDirs
